test/org/javarosa/core/TestBenchmark.java

```diff
@@ -0,0 +1,198 @@
+            //.addProfiler(WinPerfAsmProfiler.class)
+            .build();
+
+        RunResult run = new Runner(opt).run().iterator().next();
+        //run.
+    }
+
+
+    // The JMH samples are the best documentation for how to use it
+    // http://hg.openjdk.java.net/code-tools/jmh/file/tip/jmh-samples/src/main/java/org/openjdk/jmh/samples/
+    @State(Scope.Thread)
+    public static class BenchmarkState {
+        private TreeElement dataRootNode;
+        private TreeElement savedRoot;
+        private FormDef formDef;
+
+        @Setup(Level.Trial)
+        public void
+        initialize() {
+
+            FormParseInit formParseInit = new FormParseInit(r("nigeria_wards_external_combined.xml"));
+
+            FormEntryController formEntryController = formParseInit.getFormEntryController();
+
+            byte[] formInstanceAsBytes = null;
+            try {
+                formInstanceAsBytes = Files.readAllBytes(Paths.get(PathConst.getTestResourcePath().getAbsolutePath(), "populate-nodes-attributes-instance.xml"));
+            } catch (IOException e) {
+                fail("There was a problem with reading the test data.\n" + e.getMessage());
+            }
+            savedRoot = XFormParser.restoreDataModel(formInstanceAsBytes, null).getRoot();
+            formDef = formEntryController.getModel().getForm();
+            dataRootNode = formDef.getInstance().getRoot().deepCopy(true);
+        }
+    }
+
+    @State(Scope.Thread)
+    public static class ParseWithExternalInstanceState {
+        Path resourcePath = PathConst.getTestResourcePath().toPath();
+
+        Path xFormFilePath = r("nigeria_wards_external.xml");
+        FormDef formDef;
+        @Setup(Level.Trial)
+        public void
+        initialize() {
+            Path resourcePath = PathConst.getTestResourcePath().toPath();
+            ReferenceManager.instance().addReferenceFactory((buildReferenceFactory("file", resourcePath.toString())));
+
+
+        }
+    }
+
+
+    @State(Scope.Thread)
+    public static class ParseWithInternalInstanceState {
+        Path resourcePath = PathConst.getTestResourcePath().toPath();
+
+        Path xFormFilePath = r("nigeria_wards_external_combined.xml");
+        FormDef formDef;
+        @Setup(Level.Trial)
+        public void
+        initialize() {
+            Path resourcePath = PathConst.getTestResourcePath().toPath();
+            ReferenceManager.instance().addReferenceFactory((buildReferenceFactory("file", resourcePath.toString())));
+
+
+        }
+    }
+
+    @State(Scope.Thread)
+    public static class ExternalDataInstanceState {
+        ExternalDataInstance externalDataInstance = null;
+
+        @Setup(Level.Trial)
+        public void
+        initialize() {
+            Path resourcePath = PathConst.getTestResourcePath().toPath();
+            ReferenceManager.instance().addReferenceFactory((buildReferenceFactory("file", resourcePath.toString())));
```

Comment:
Since ReferenceManager is a singleton, better use `ReferenceManagerTestUtils.setUpSimpleReferenceManager()` instead to avoid some sources of interference between across different benchmarks.